beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,29 +1729,85 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         let parent_root = block.parent_root;
         let slot = block.slot;
 
+        // Clone the state, without the tree hash cache (THC), then "rename" the original `state` (with
+        // the THC) to `pre_state`. After this we have:
+        //
+        // - `state`: does not contain THC.
+        // - `pre_state`: contains THC 🍁.
+        //
+        // The reason for this clone is explained later in this function.
+        let cloned_state = state.clone_with(CloneConfig::committee_caches_only());
+        let mut pre_state = std::mem::replace(&mut state, cloned_state);
+
+        // Advance a single slot on `pre_state`.
+        //
+        // This is an optimisation with the following benefits:
+        //
+        // 1. When following head (i.e., not syncing), this means that we can use the tail-end of
+        //    slot `n` to compute the pre-state for the block at slot `n + 1`. This allows us to
+        //    import blocks faster ("import" meaning placed in the DB and fork choice, not
+        //    necessarily the completion of this function).
+        // 2. On epoch boundaries, it allows us to learn the proposer shuffling for the next epoch
+        //    and prime our caches. This shortens block propagation verification times, since we
+        //    don't need to compute the shuffling.
+        //
+        // The downside of this optimization is that we now need to hold two copies of the state;
+        // one that is advanced to the next state (`pre_state`) and one that is not (`state`). We
+        // maintain the non-advanced `state` to avoid a DB read when setting an imported block as
+        // the head and therefore putting the state in `self.canonical_head`.
+        if let Some(summary) =
+            per_slot_processing(&mut pre_state, Some(block.state_root), &self.spec)?
```

Comment:
I need to think about this some more, but I'll write something now before I come back to it tomorrow...

I can see how computing the proposers at the end of the epoch takes pressure off the gossip verification code at the start of the epoch (yay), but don't we also shift the burden of epoch processing onto the last slot? I.e. is there any risk that the cost of epoch processing might cause us to miss head votes on the last slot of the epoch -- just as we sometimes miss them on the first slot of the epoch right now?

When I imagined slot processing happening in the last slot, I thought it might happen as an independent step from block import, which would avoid the issue. Perhaps you've already done the hard work of thinking through the pros and cons of this alternative? Having a timer that fires and triggers slot processing might be a bit nasty?

Reply:
Oh yeah, this is a good point. My reasoning was that we've already placed the block in the DB/FC so it's already "imported", but we still haven't updated the head so we're at risk of missing attestations to the head... This is something I'll have to think about, thanks :)